File: results/2026-07-28_114218/tasks/11-semantic-version-bumper/default-haiku45/generated-code/version_bumper.py
"""
Semantic version bumper: parse version, determine bump from commits, update files.

Implements conventional commit parsing (feat, fix, BREAKING CHANGE) to determine
the next semantic version (major.minor.patch) and generates changelog entries.
"""

import json
import re
from pathlib import Path
from typing import Optional


def parse_version(file_path: str) -> str:
    """
    Parse semantic version from a file.

    Supports:
    - package.json with 'version' field
    - Plain text VERSION file with version string
    - Versions with optional 'v' prefix

    Args:
        file_path: Path to version file

    Returns:
        Semantic version string (e.g., '1.2.3')

    Raises:
        FileNotFoundError: If file does not exist
        ValueError: If version cannot be parsed
    """
    path = Path(file_path)
    if not path.exists():
        raise FileNotFoundError(f'Version file not found: {file_path}')

    content = path.read_text().strip()

    # Try JSON first
    if file_path.endswith('.json'):
        try:
            data = json.loads(content)
            version = data.get('version', '').strip()
            if version:
                return version.lstrip('v')
        except json.JSONDecodeError:
            pass

    # Plain text: first line
    version = content.split('\n')[0].strip()
    if not version:
        raise ValueError(f'No version found in {file_path}')

    return version.lstrip('v')


def determine_bump_type(commit_message: str) -> Optional[str]:
    """
    Determine version bump type from a commit message.

    Implements conventional commits:
    - BREAKING CHANGE (anywhere): major bump
    - feat: minor bump
    - fix: patch bump
    - other (chore, docs, style, etc.): no bump

    Args:
        commit_message: Full commit message (subject + body)

    Returns:
        'major', 'minor', 'patch', or None (no bump)
    """
    # Check for breaking change first (highest priority)
    if 'BREAKING CHANGE' in commit_message or 'BREAKING-CHANGE' in commit_message:
        return 'major'

    # Check commit type in subject line (first line)
    subject = commit_message.split('\n')[0].lower()

    if subject.startswith('feat'):
        return 'minor'
    if subject.startswith('fix'):
        return 'patch'

    return None


def get_next_version(current: str, bump_type: Optional[str]) -> str:
    """
    Calculate next semantic version based on bump type.

    Semantic versioning: major.minor.patch
    - Major bump: major += 1, minor = 0, patch = 0
    - Minor bump: minor += 1, patch = 0
    - Patch bump: patch += 1
    - No bump: version unchanged

    Args:
        current: Current version (e.g., '1.2.3')
        bump_type: 'major', 'minor', 'patch', or None

    Returns:
        Next version string

    Raises:
        ValueError: If current version is not valid semver
    """
    # Validate semver format
    match = re.match(r'^(\d+)\.(\d+)\.(\d+)$', current)
    if not match:
        raise ValueError(f'Invalid semantic version format: {current}')

    major, minor, patch = map(int, match.groups())

    if bump_type is None:
        return current
    elif bump_type == 'major':
        return f'{major + 1}.0.0'
    elif bump_type == 'minor':
        return f'{major}.{minor + 1}.0'
    elif bump_type == 'patch':
        return f'{major}.{minor}.{patch + 1}'
    else:
        raise ValueError(f'Invalid bump type: {bump_type}')


def update_version_file(file_path: str, new_version: str) -> None:
    """
    Update version in a file.

    Supports:
    - package.json: updates 'version' field
    - Plain text VERSION file: replaces entire content

    Args:
        file_path: Path to version file
        new_version: New version string

    Raises:
        FileNotFoundError: If file does not exist
    """
    path = Path(file_path)
    if not path.exists():
        raise FileNotFoundError(f'Version file not found: {file_path}')

    if file_path.endswith('.json'):
        data = json.loads(path.read_text())
        data['version'] = new_version
        path.write_text(json.dumps(data, indent=2))
    else:
        path.write_text(new_version)


def generate_changelog_entry(commits: list[str], version: str) -> str:
    """
    Generate a changelog entry from commit messages.

    Formats commits by type (breaking changes, features, fixes) in markdown.

    Args:
        commits: List of commit messages
        version: Version number for this release

    Returns:
        Markdown-formatted changelog entry
    """
    breaking = []
    features = []
    fixes = []

    for commit in commits:
        subject = commit.split('\n')[0]

        if 'BREAKING CHANGE' in commit or 'BREAKING-CHANGE' in commit:
            breaking.append(subject)
        elif subject.startswith('feat'):
            features.append(subject)
        elif subject.startswith('fix'):
            fixes.append(subject)

    # Format as markdown
    lines = [f'## {version}\n']

    if breaking:
        lines.append('### 🔴 Breaking Changes')
        for msg in breaking:
            lines.append(f'- {msg}')
        lines.append('')

    if features:
        lines.append('### ✨ Features')
        for msg in features:
            lines.append(f'- {msg}')
        lines.append('')

    if fixes:
        lines.append('### 🐛 Fixes')
        for msg in fixes:
            lines.append(f'- {msg}')
        lines.append('')

    return '\n'.join(lines)

File: results/2026-07-28_205523/tasks/11-semantic-version-bumper/default-haiku45/generated-code/version-bumper.js
// Semantic Version Bumper
// Follows conventional commits and semantic versioning
// Uses TDD approach: minimal implementation for each test

const fs = require('fs');
const path = require('path');

// Parse semantic version from package.json or VERSION file
function parseVersion(filePath = 'package.json') {
  if (!fs.existsSync(filePath)) {
    throw new Error(`Version file not found: ${filePath}`);
  }

  if (filePath.endsWith('.json')) {
    const content = JSON.parse(fs.readFileSync(filePath, 'utf8'));
    if (!content.version) {
      throw new Error('No version field in ' + filePath);
    }
    return content.version;
  } else {
    return fs.readFileSync(filePath, 'utf8').trim();
  }
}

// Determine version bump type based on conventional commits
// Priority: breaking > minor > patch
function determineVersionBump(commits) {
  let hasMajor = false;
  let hasMinor = false;
  let hasPatch = false;

  for (const commit of commits) {
    const message = commit.message || '';

    // Check for breaking changes
    if (message.includes('BREAKING CHANGE:') || message.startsWith('BREAKING CHANGE:')) {
      hasMajor = true;
    }

    // Check commit type
    if (message.startsWith('feat:') || message.startsWith('feat(')) {
      hasMinor = true;
    } else if (message.startsWith('fix:') || message.startsWith('fix(')) {
      hasPatch = true;
    }
  }

  if (hasMajor) return 'major';
  if (hasMinor) return 'minor';
  if (hasPatch) return 'patch';

  return 'patch'; // Default to patch if no conventional commits found
}

// Calculate next version based on current version and bump type
function calculateNextVersion(currentVersion, bumpType) {
  const parts = currentVersion.split('.').map(Number);
  const [major, minor, patch] = parts;

  if (bumpType === 'major') {
    return `${major + 1}.0.0`;
  } else if (bumpType === 'minor') {
    return `${major}.${minor + 1}.0`;
  } else if (bumpType === 'patch') {
    return `${major}.${minor}.${patch + 1}`;
  }

  throw new Error(`Unknown bump type: ${bumpType}`);
}

// Update version in file (package.json or VERSION)
function updateVersion(newVersion, filePath = 'package.json') {
  if (filePath.endsWith('.json')) {
    const content = JSON.parse(fs.readFileSync(filePath, 'utf8'));
    content.version = newVersion;
    fs.writeFileSync(filePath, JSON.stringify(content, null, 2) + '\n', 'utf8');
  } else {
    fs.writeFileSync(filePath, newVersion + '\n', 'utf8');
  }
}

// Generate changelog entry from commits
function generateChangelogEntry(version, commits) {
  const date = new Date().toISOString().split('T')[0];
  const lines = [`## [${version}] - ${date}`];

  // Categorize commits
  const features = [];
  const fixes = [];
  const breaking = [];

  for (const commit of commits) {
    const message = commit.message || '';

    // Extract breaking changes
    if (message.includes('BREAKING CHANGE:') || message.startsWith('BREAKING CHANGE:')) {
      const match = message.match(/BREAKING CHANGE:\s*(.+?)(?:\n|$)/);
      if (match) {
        breaking.push(match[1].trim());
      }
    }

    // Categorize by type
    if (message.startsWith('feat:') || message.startsWith('feat(')) {
      const description = message.replace(/^feat\(\w+\):\s*/, '').replace(/^feat:\s*/, '');
      features.push(description.split('\n')[0]);
    } else if (message.startsWith('fix:') || message.startsWith('fix(')) {
      const description = message.replace(/^fix\(\w+\):\s*/, '').replace(/^fix:\s*/, '');
      fixes.push(description.split('\n')[0]);
    }
  }

  // Format sections
  if (breaking.length > 0) {
    lines.push('');
    lines.push('### BREAKING CHANGES');
    breaking.forEach(b => lines.push(`- ${b}`));
  }

  if (features.length > 0) {
    lines.push('');
    lines.push('### Features');
    features.forEach(f => lines.push(`- ${f}`));
  }

  if (fixes.length > 0) {
    lines.push('');
    lines.push('### Bug Fixes');
    fixes.forEach(fix => lines.push(`- ${fix}`));
  }

  return lines.join('\n');
}

// Main function: run the complete version bumper
function runVersionBumper(commits, versionFile = 'package.json') {
  // Parse current version
  const currentVersion = parseVersion(versionFile);

  // Determine bump type from commits
  const bumpType = determineVersionBump(commits);

  // Calculate next version
  const newVersion = calculateNextVersion(currentVersion, bumpType);

  // Update version file
  updateVersion(newVersion, versionFile);

  // Generate changelog
  const changelog = generateChangelogEntry(newVersion, commits);

  return {
    currentVersion,
    newVersion,
    bumpType,
    changelog
  };
}

module.exports = {
  parseVersion,
  determineVersionBump,
  calculateNextVersion,
  updateVersion,
  generateChangelogEntry,
  runVersionBumper
};
